### PR TITLE
Build and assemble manifest with lock.

### PR DESCRIPTION
**Pull request by dblock**
Signed-off-by: dblock <dblock@amazon.com>

### Description

- Builds and assembles manifest with lock and uploads the lock SHA to S3. Not yet aborting a build if there were no changes, but gets us _this_ close.
- Consolidates testing under a `dryRun` argument, gets rid of `upload` and `script`.

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: tests/jenkins/TestVars.groovy [hunk 1]
```diff
@@ -30,6 +30,10 @@ class TestVars extends BuildPipelineTest {
 
         binding.setVariable('WEBHOOK_URL', 'https://web/hook/url')
 
+        helper.registerAllowedMethod("sha1", [String], { filaname ->
```
Comment (owaiskazi19) [inline tests/jenkins/TestVars.groovy:33]: Should it be `filename`?

File: vars/buildAndAssemble.groovy [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    String manifestLock = args.dryRun ? manifest : "${manifest}.lock"
+    manifestSHA = sha1(manifestLock)
+    echo "Manifest SHA: ${manifestSHA}"
+
+    lib = library(identifier: 'jenkins@20211122', retriever: legacySCM(scm))
+    def inputManifest = lib.jenkins.InputManifest.new(readYaml(file: manifestLock))
+    String shasRoot = inputManifest.getSHAsRoot("${JOB_NAME}", args.platform, args.architecture)
+
+    buildManifest(
+        args + [
```
Comment (owaiskazi19) [inline vars/buildAndAssemble.groovy:20]: Can use `args.extend([manifest: manifestLock])`

File: tests/jenkins/TestVars.groovy (hunk 1 above)
Comment (dblock) [reply tests/jenkins/TestVars.groovy]: fixed

File: vars/buildAndAssemble.groovy (hunk 2 above)
Comment (dblock) [reply vars/buildAndAssemble.groovy:20]: Isn't that in-place? Either way it doesn't seem any better, or is it?

Comment (owaiskazi19) [reply vars/buildAndAssemble.groovy:20]: Actually `+` creates a new list and `extends` add it in the existing list. There's a good article for the comparison https://blog.finxter.com/python-list-concatenation-add-vs-inplace-add-vs-extend/

Comment (dblock) [reply vars/buildAndAssemble.groovy:20]: Right, I do not want to modify the value being passed in.

**Review by owaiskazi19 — APPROVED** [review]
LGTM

File: jenkins/opensearch-dashboards/Jenkinsfile [hunk 3]
```diff
@@ -18,7 +18,7 @@ pipeline {
         buildDiscarder(logRotator(artifactNumToKeepStr: '1'))
     }
     stages {
-        stage('detect Docker image + args to use for the build') {
+        stage('detect') {
```
Comment (gaiksaya) [inline jenkins/opensearch-dashboards/Jenkinsfile:21]: Can you say more? Like `Detect CI image` or something else. Just detect is confusing for external folks

File: jenkins/opensearch/Jenkinsfile [hunk 4]
```diff
@@ -16,7 +16,7 @@ pipeline {
         )
     }
     stages {
-        stage('detect Docker image + args to use for the build') {
+        stage('detect') {
```
Comment (gaiksaya) [inline jenkins/opensearch/Jenkinsfile:19]: Same as above

File: jenkins/opensearch/Jenkinsfile [hunk 5]
```diff
@@ -77,9 +76,7 @@ pipeline {
                     }
                     steps {
                         script {
-                            buildManifest()
-                            assembleManifest()
-                            uploadArtifacts(upload: true)
+                            buildAndAssemble()
```
Comment (gaiksaya) [inline jenkins/opensearch/Jenkinsfile:79]: Should the upload step be different? Not sure if we always want to build and upload especially for someone experimenting with the script. Or change the method name to `buildAndAssembleAndUpload` just to give headsup that this method uploads too

File: vars/buildManifest.groovy [hunk 6]
```diff
@@ -1,9 +1,10 @@
 void call(Map args = [:]) {
     sh (([
-        args.script ?: './build.sh',
+        args.dryRun ? 'echo ./build.sh' : './build.sh',
```
Comment (gaiksaya) [inline vars/buildManifest.groovy:3]: Did we need to modify the build.sh to add `dryRun` arg or somewhere else?

File: jenkins/opensearch-dashboards/Jenkinsfile (hunk 3 above)
Comment (dblock) [reply jenkins/opensearch-dashboards/Jenkinsfile:21]: I really just wanted to lowercase and shorten it, changed to `detect docker image + args`.

File: jenkins/opensearch/Jenkinsfile (hunk 5 above)
Comment (dblock) [reply jenkins/opensearch/Jenkinsfile:79]: Renamed to `buildAssembleUpload`

File: vars/buildManifest.groovy (hunk 6 above)
Comment (dblock) [reply vars/buildManifest.groovy:3]: In this case "dry run" = "replace command with echo command ..." but yes, we could extend `build.sh`, `assemble.sh` and `test.sh` to take `--dry-run`. I'm not sure it's much better though.

**Review by gaiksaya — APPROVED** [review]
LGTM

**Pull request merged**